mypy/meet.py

```diff
@@ -51,155 +56,246 @@ def narrow_declared_type(declared: Type, narrowed: Type) -> Type:
-        num_same = len(t_set.intersection(s_set))
-        num_diff = len(t_set.symmetric_difference(s_set))
-        return num_same > 0 and num_diff > 0
-
-    # Are tuples partially overlapping?
-    tup_overlap = is_overlapping_tuples(t, s, use_promotions=True)
-    if tup_overlap is not None and tup_overlap:
-        return tup_overlap
-
-    def is_object(t: Type) -> bool:
-        return isinstance(t, Instance) and t.type.fullname() == 'builtins.object'
-
-    # Is either 't' or 's' an unrestricted TypeVar?
-    if isinstance(t, TypeVarType) and is_object(t.upper_bound) and len(t.values) == 0:
-        return True
+def get_possible_variants(typ: Type) -> List[Type]:
+    """This function takes any "Union-like" type and returns a list of the available "options".
 
-    if isinstance(s, TypeVarType) and is_object(s.upper_bound) and len(s.values) == 0:
-        return True
+    Specifically, there are currently exactly three different types that can have
+    "variants" or are "union-like":
 
-    return False
+    - Unions
+    - TypeVars with value restrictions
+    - Overloads
 
+    This function will return a list of each "option" present in those types.
 
-def is_overlapping_types(t: Type, s: Type, use_promotions: bool = False) -> bool:
-    """Can a value of type t be a value of type s, or vice versa?
+    If this function receives any other type, we return a list containing just that
+    original type. (E.g. pretend the type was contained within a singleton union).
 
-    Note that this effectively checks against erased types, since type
-    variables are erased at runtime and the overlapping check is based
-    on runtime behavior. The exception is protocol types, it is not safe,
-    but convenient and is an opt-in behavior.
+    The only exception is regular TypeVars: we return a list containing that TypeVar's
+    upper bound.
 
-    If use_promotions is True, also consider type promotions (int and
-    float would only be overlapping if it's True).
+    This function is useful primarily when checking to see if two types are overlapping:
+    the algorithm to check if two unions are overlapping is fundamentally the same as
+    the algorithm for checking if two overloads are overlapping.
 
-    This does not consider multiple inheritance. For example, A and B in
-    the following example are not considered overlapping, even though
-    via C they can be overlapping:
+    Normalizing both kinds of types in the same way lets us reuse the same algorithm
+    for both.
+    """
+    if isinstance(typ, TypeVarType):
+        if len(typ.values) > 0:
+            return typ.values
+        else:
+            return [typ.upper_bound]
+    elif isinstance(typ, UnionType):
+        return typ.relevant_items()
+    elif isinstance(typ, Overloaded):
+        # Note: doing 'return typ.items()' makes mypy
+        # infer a too-specific return type of List[CallableType]
+        out = []  # type: List[Type]
+        out.extend(typ.items())
```

Comment:
Just use `return list(typ.iems())` (i.e. make a copy so that mypy is happy about invariance).